// @mintlify/cli/bin/workflows/client.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { authenticatedFetch } from '../authenticatedFetch.js';
import { API_URL } from '../constants.js';
function request(path, init) {
    return __awaiter(this, void 0, void 0, function* () {
        const res = yield authenticatedFetch(`${API_URL}/api/cli/workflows${path}`, Object.assign(Object.assign({}, init), { headers: Object.assign({ Accept: 'application/json' }, init === null || init === void 0 ? void 0 : init.headers) }));
        if (!res.ok) {
            const body = yield res.text().catch(() => '');
            throw new Error(`API error (${res.status}): ${body || res.statusText}`);
        }
        return res.json();
    });
}
export function createWorkflow(subdomain, body) {
    return request(`/${encodeURIComponent(subdomain)}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
    });
}
export function listWorkflows(subdomain) {
    return request(`/${encodeURIComponent(subdomain)}`);
}
export function deleteWorkflow(subdomain, workflowSchemaId) {
    return request(`/${encodeURIComponent(subdomain)}/${encodeURIComponent(workflowSchemaId)}`, { method: 'DELETE' });
}
