// @mintlify/cli/bin/workflows/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { jsx as _jsx } from "react/jsx-runtime";
import { addLog, ErrorLog, SpinnerLog, SuccessLog, removeLastLog } from '@mintlify/previewing';
import chalk from 'chalk';
import { Text } from 'ink';
import yaml from 'js-yaml';
import fs from 'node:fs/promises';
import path from 'node:path';
import { CMD_EXEC_PATH, terminate } from '../helpers.js';
import { subdomainMiddleware } from '../middlewares/subdomainMiddleware.js';
import { trackEvent } from '../telemetry/track.js';
import { createWorkflow, deleteWorkflow, listWorkflows } from './client.js';
import { WORKFLOW_TYPES, } from './types.js';
function resolveFormat(argv) {
    return argv.format === 'json' ? 'json' : 'table';
}
function output(format, text) {
    if (format === 'table') {
        addLog(_jsx(Text, { children: text }));
    }
    else {
        process.stdout.write(text + '\n');
    }
}
function requireSubdomain(subdomain) {
    if (!subdomain) {
        throw new Error('No subdomain set. Pass --subdomain, or run `mint config set subdomain <name>`.');
    }
    return subdomain;
}
function readWorkflowFile(filePath) {
    return __awaiter(this, void 0, void 0, function* () {
        const resolved = path.isAbsolute(filePath) ? filePath : path.join(CMD_EXEC_PATH, filePath);
        const contents = yield fs.readFile(resolved, 'utf-8');
        if (/\.ya?ml$/i.test(resolved))
            return yaml.load(contents);
        return JSON.parse(contents);
    });
}
function buildTriggerFromFlags(argv) {
    var _a;
    const repos = ((_a = argv.pushRepo) !== null && _a !== void 0 ? _a : []).map(String).filter(Boolean);
    if (argv.cron && repos.length > 0) {
        throw new Error('Specify either --cron or --push-repo, not both.');
    }
    if (argv.cron)
        return { cron: argv.cron };
    if (repos.length > 0)
        return { push: repos.map((repo) => ({ repo })) };
    throw new Error('A trigger is required. Pass --cron <expr> or --push-repo <owner/repo>.');
}
function buildBodyFromFlags(argv) {
    var _a;
    if (!argv.name) {
        throw new Error('--name is required when --file is not provided.');
    }
    const type = argv.type;
    if (type !== undefined && !WORKFLOW_TYPES.includes(type)) {
        throw new Error(`Invalid --type "${type}". Valid types: ${WORKFLOW_TYPES.join(', ')}`);
    }
    const body = {
        name: argv.name,
        on: buildTriggerFromFlags(argv),
    };
    if (argv.prompt)
        body.prompt = argv.prompt;
    if (type)
        body.type = type;
    if (argv.automerge !== undefined)
        body.automerge = argv.automerge;
    const contextRepos = ((_a = argv.contextRepo) !== null && _a !== void 0 ? _a : []).map(String).filter(Boolean);
    if (contextRepos.length > 0)
        body.context = contextRepos.map((repo) => ({ repo }));
    return body;
}
function describeTrigger(trigger) {
    if ('cron' in trigger)
        return `cron(${trigger.cron})`;
    if ('push' in trigger)
        return `push(${trigger.push.map((p) => p.repo).join(', ')})`;
    return `composio(${trigger.composio.toolkit}:${trigger.composio.triggerSlug})`;
}
function renderWorkflow(workflow) {
    const lines = [];
    lines.push(chalk.bold(`\nWorkflow ${workflow.name}`));
    lines.push(`  ${chalk.dim('ID')}        ${workflow._id}`);
    lines.push(`  ${chalk.dim('Status')}    ${workflow.status}`);
    if (workflow.type)
        lines.push(`  ${chalk.dim('Type')}      ${workflow.type}`);
    lines.push(`  ${chalk.dim('Trigger')}   ${describeTrigger(workflow.trigger)}`);
    if (workflow.automerge !== undefined) {
        lines.push(`  ${chalk.dim('Automerge')} ${workflow.automerge}`);
    }
    return lines.join('\n');
}
function renderList(workflows) {
    if (workflows.length === 0)
        return 'No workflows found.';
    const rows = workflows.map((w) => {
        var _a;
        return [
            w._id,
            w.name,
            (_a = w.type) !== null && _a !== void 0 ? _a : '—',
            describeTrigger(w.trigger),
            w.status,
        ];
    });
    const headers = ['ID', 'Name', 'Type', 'Trigger', 'Status'];
    const widths = headers.map((h, i) => Math.max(h.length, ...rows.map((r) => String(r[i]).length)));
    const fmt = (cells) => cells.map((c, i) => c.padEnd(widths[i])).join('  ');
    return [chalk.bold(fmt(headers)), ...rows.map(fmt)].join('\n');
}
const withSubdomain = (yargs) => yargs.option('subdomain', {
    type: 'string',
    description: 'Documentation subdomain (default: mint config set subdomain)',
});
const withFormat = (yargs) => yargs.option('format', {
    type: 'string',
    choices: ['table', 'json'],
    description: 'Output format (table=pretty, json=raw)',
});
export const workflowsBuilder = (yargs) => yargs
    .middleware(subdomainMiddleware)
    .command('create', 'Create a new workflow', (yargs) => withFormat(withSubdomain(yargs)
    .option('name', { type: 'string', description: 'Workflow name' })
    .option('prompt', { type: 'string', description: 'Workflow prompt' })
    .option('type', {
    type: 'string',
    choices: WORKFLOW_TYPES,
    description: 'Workflow type',
})
    .option('cron', {
    type: 'string',
    description: 'Cron expression for a scheduled trigger',
})
    .option('push-repo', {
    type: 'string',
    array: true,
    description: 'Repository (owner/repo) for a push trigger (repeatable)',
})
    .option('context-repo', {
    type: 'string',
    array: true,
    description: 'Additional context repository (repeatable)',
})
    .option('automerge', {
    type: 'boolean',
    description: 'Automatically merge PRs opened by this workflow',
})
    .option('file', {
    type: 'string',
    description: 'Path to a JSON or YAML file with the full workflow body',
}))
    .example('mint workflow create --name Translations --cron "0 6 * * *" --type translations', '')
    .example('mint workflow create --file workflow.yaml', ''), (argv) => __awaiter(void 0, void 0, void 0, function* () {
    const format = resolveFormat(argv);
    try {
        const subdomain = requireSubdomain(argv.subdomain);
        const body = argv.file
            ? (yield readWorkflowFile(argv.file))
            : buildBodyFromFlags(argv);
        if (format === 'table')
            addLog(_jsx(SpinnerLog, { message: "Creating workflow..." }));
        const created = yield createWorkflow(subdomain, body);
        if (format === 'table')
            removeLastLog();
        void trackEvent('cli.workflow.created', {
            subdomain,
            type: created.type,
            triggerKind: describeTrigger(created.trigger).split('(')[0],
        });
        if (format === 'json') {
            output(format, JSON.stringify(created, null, 2));
        }
        else {
            addLog(_jsx(SuccessLog, { message: `Created workflow ${created._id}` }));
            output(format, renderWorkflow(created));
        }
        yield terminate(0);
    }
    catch (err) {
        if (format === 'table')
            removeLastLog();
        const message = err instanceof Error ? err.message : 'unknown error';
        if (format === 'json') {
            process.stderr.write(`Error: ${message}\n`);
        }
        else {
            addLog(_jsx(ErrorLog, { message: message }));
        }
        yield terminate(1);
    }
}))
    .command('list', 'List workflows for the current deployment', (yargs) => withFormat(withSubdomain(yargs)), (argv) => __awaiter(void 0, void 0, void 0, function* () {
    const format = resolveFormat(argv);
    try {
        const subdomain = requireSubdomain(argv.subdomain);
        if (format === 'table')
            addLog(_jsx(SpinnerLog, { message: "Fetching workflows..." }));
        const workflows = yield listWorkflows(subdomain);
        if (format === 'table')
            removeLastLog();
        if (format === 'json') {
            output(format, JSON.stringify(workflows, null, 2));
        }
        else {
            output(format, renderList(workflows));
        }
        yield terminate(0);
    }
    catch (err) {
        if (format === 'table')
            removeLastLog();
        const message = err instanceof Error ? err.message : 'unknown error';
        if (format === 'json') {
            process.stderr.write(`Error: ${message}\n`);
        }
        else {
            addLog(_jsx(ErrorLog, { message: message }));
        }
        yield terminate(1);
    }
}))
    .command('delete <id>', 'Delete a workflow by ID', (yargs) => withFormat(withSubdomain(yargs)).positional('id', {
    type: 'string',
    demandOption: true,
    description: 'Workflow schema ID',
}), (argv) => __awaiter(void 0, void 0, void 0, function* () {
    const format = resolveFormat(argv);
    try {
        const subdomain = requireSubdomain(argv.subdomain);
        if (format === 'table')
            addLog(_jsx(SpinnerLog, { message: "Deleting workflow..." }));
        yield deleteWorkflow(subdomain, argv.id);
        if (format === 'table')
            removeLastLog();
        if (format === 'json') {
            output(format, JSON.stringify({ success: true, id: argv.id }, null, 2));
        }
        else {
            addLog(_jsx(SuccessLog, { message: `Deleted workflow ${argv.id}` }));
        }
        yield terminate(0);
    }
    catch (err) {
        if (format === 'table')
            removeLastLog();
        const message = err instanceof Error ? err.message : 'unknown error';
        if (format === 'json') {
            process.stderr.write(`Error: ${message}\n`);
        }
        else {
            addLog(_jsx(ErrorLog, { message: message }));
        }
        yield terminate(1);
    }
}))
    .demandCommand(1, 'specify a subcommand: create, list, or delete');
