// @mintlify/models/dist/mintconfig/localization.js
/*
Currently supported languages, should be ISO 639:
English - en
Chinese - cn, ISO 639 = zh
Simplified Chinese - zh-Hans
Traditional Chinese - zh-Hant
Spanish - es
French - fr
Japanese - jp, ISO 639 = ja
Portuguese - pt
Brazilian Portugese - pt-BR
German - de
Korean - ko
Italian - it
Russian - ru
Romanian - ro
Czech - cs
Indonesian - id
Arabic - ar
Turkish - tr
Hindi - hi
Swedish - sv
Norwegian - no
Danish - da
Latvian - lv
Dutch - nl
Ukrainian - uk
Vietnamese - vi
Polish - pl
Uzbek - uz
Hebrew - he
Catalan - ca
Hungarian - hu
Finnish - fi
*/
export const locales = [
    'en',
    'cn',
    'zh',
    'zh-Hans',
    'zh-Hant',
    'es',
    'fr',
    'fr-CA',
    'fr-ca',
    'ja',
    'jp',
    'ja-jp',
    'pt',
    'pt-BR',
    'de',
    'ko',
    'it',
    'ru',
    'ro',
    'cs',
    'id',
    'ar',
    'tr',
    'hi',
    'sv',
    'no',
    'da',
    'lv',
    'nl',
    'uk',
    'vi',
    'pl',
    'uz',
    'he',
    'ca',
    'fi',
    'hu',
];
export const localeDisplayNames = {
    en: 'English',
    cn: 'Chinese',
    zh: 'Chinese',
    'zh-Hans': 'Simplified Chinese',
    'zh-Hant': 'Traditional Chinese',
    es: 'Spanish',
    fr: 'French',
    'fr-CA': 'Canadian French',
    'fr-ca': 'Canadian French',
    ja: 'Japanese',
    jp: 'Japanese',
    'ja-jp': 'Japanese',
    pt: 'Portuguese',
    'pt-BR': 'Brazilian Portuguese',
    de: 'German',
    ko: 'Korean',
    it: 'Italian',
    ru: 'Russian',
    ro: 'Romanian',
    cs: 'Czech',
    id: 'Indonesian',
    ar: 'Arabic',
    tr: 'Turkish',
    hi: 'Hindi',
    sv: 'Swedish',
    no: 'Norwegian',
    da: 'Danish',
    lv: 'Latvian',
    nl: 'Dutch',
    uk: 'Ukrainian',
    vi: 'Vietnamese',
    pl: 'Polish',
    uz: 'Uzbek',
    he: 'Hebrew',
    ca: 'Catalan',
    fi: 'Finnish',
    hu: 'Hungarian',
};
export const localeDisplayFlags = {
    en: '馃嚭馃嚫',
    cn: '馃嚚馃嚦',
    zh: '馃嚚馃嚦',
    'zh-Hans': '馃嚚馃嚦',
    'zh-Hant': '馃嚬馃嚰',
    es: '馃嚜馃嚫',
    fr: '馃嚝馃嚪',
    'fr-CA': '馃嚚馃嚘',
    'fr-ca': '馃嚚馃嚘',
    ja: '馃嚡馃嚨',
    jp: '馃嚡馃嚨',
    'ja-jp': '馃嚚馃嚘',
    pt: '馃嚙馃嚪',
    'pt-BR': '馃嚙馃嚪',
    de: '馃嚛馃嚜',
    ko: '馃嚢馃嚪',
    it: '馃嚠馃嚬',
    ru: '馃嚪馃嚭',
    ro: '馃嚪馃嚧',
    cs: '馃嚚馃嚳',
    id: '馃嚠馃嚛',
    ar: '馃嚫馃嚘',
    tr: '馃嚬馃嚪',
    hi: '馃嚠馃嚦',
    sv: '馃嚫馃嚜',
    no: '馃嚦馃嚧',
    da: '馃嚛馃嚢',
    lv: '馃嚤馃嚮',
    nl: '馃嚦馃嚤',
    uk: '馃嚭馃嚘',
    vi: '馃嚮馃嚦',
    pl: '馃嚨馃嚤',
    uz: '馃嚭馃嚳',
    he: '馃嚠馃嚤',
    ca: '馃嚜馃嚫',
    fi: '馃嚝馃嚠',
    hu: '馃嚟馃嚭',
};
export const getLocaleDisplayName = (locale) => {
    return localeDisplayNames[locale];
};
export const getLocaleDisplayFlag = (locale) => {
    return localeDisplayFlags[locale];
};
export const LocaleCodes = Object.keys(localeDisplayNames);
export const LocaleDisplayNames = Object.values(localeDisplayNames);
export const LocaleDisplayFlags = Object.values(localeDisplayFlags);
export const LlmSupportedLocaleCodes = LocaleCodes;
export const LlmSupportedLocaleDisplayNames = LocaleDisplayNames;
export const LlmSupportedLocaleDisplayFlags = LocaleDisplayFlags;
