// @mintlify/models/dist/types/creditTier.js
export const CreditTierIds = [
    'tier-0',
    'tier-5000',
    'tier-10250',
    'tier-15250',
    'tier-26000',
    'tier-31000',
    'tier-52500',
    'tier-57500',
    'tier-80000',
    'tier-85000',
    'tier-108500',
    'tier-113500',
    'tier-275000',
    'tier-280000',
    'tier-555000',
    'tier-560000',
    'tier-850000',
    'tier-855000',
    'tier-1150000',
    'tier-1155000',
];
export const CREDIT_TIER_OVERAGE_PRICE_PER_CREDIT = 0.01;
export const CREDIT_TIERS = {
    'tier-0': {
        name: '0',
        credits: 0,
        monthlyPrice: 0,
        validPlans: ['hobby'],
    },
    'tier-5000': {
        name: '5,000',
        credits: 5000,
        monthlyPrice: 0,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-10250': {
        name: '10,250',
        credits: 10250,
        monthlyPrice: 100,
        validPlans: ['hobby'],
    },
    'tier-15250': {
        name: '15,250',
        credits: 15250,
        monthlyPrice: 100,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-26000': {
        name: '26,000',
        credits: 26000,
        monthlyPrice: 250,
        validPlans: ['hobby'],
    },
    'tier-31000': {
        name: '31,000',
        credits: 31000,
        monthlyPrice: 250,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-52500': {
        name: '52,500',
        credits: 52500,
        monthlyPrice: 500,
        validPlans: ['hobby'],
    },
    'tier-57500': {
        name: '57,500',
        credits: 57500,
        monthlyPrice: 500,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-80000': {
        name: '80,000',
        credits: 80000,
        monthlyPrice: 750,
        validPlans: ['hobby'],
    },
    'tier-85000': {
        name: '85,000',
        credits: 85000,
        monthlyPrice: 750,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-108500': {
        name: '108,500',
        credits: 108500,
        monthlyPrice: 1000,
        validPlans: ['hobby'],
    },
    'tier-113500': {
        name: '113,500',
        credits: 113500,
        monthlyPrice: 1000,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-275000': {
        name: '275,000',
        credits: 275000,
        monthlyPrice: 2500,
        validPlans: ['hobby'],
    },
    'tier-280000': {
        name: '280,000',
        credits: 280000,
        monthlyPrice: 2500,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-555000': {
        name: '555,000',
        credits: 555000,
        monthlyPrice: 5000,
        validPlans: ['hobby'],
    },
    'tier-560000': {
        name: '560,000',
        credits: 560000,
        monthlyPrice: 5000,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-850000': {
        name: '850,000',
        credits: 850000,
        monthlyPrice: 7500,
        validPlans: ['hobby'],
    },
    'tier-855000': {
        name: '855,000',
        credits: 855000,
        monthlyPrice: 7500,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
    'tier-1150000': {
        name: '1,150,000',
        credits: 1150000,
        monthlyPrice: 10000,
        validPlans: ['hobby'],
    },
    'tier-1155000': {
        name: '1,155,000',
        credits: 1155000,
        monthlyPrice: 10000,
        validPlans: ['pro', 'proJul25', 'growth', 'enterprise'],
    },
};
