// @mintlify/validation/dist/mint-config/schemas/v2/properties/fonts/index.js
import { fontFormats } from '@mintlify/models';
import { z } from 'zod';
import { fontSourceSchema } from './source.js';
export const fontDetailsSchema = z
    .object({
    family: z
        .string()
        .describe('The font family name. Any [Google Fonts](https://fonts.google.com/) family works out of the box (e.g. `Inter`, `Open Sans`, `Playfair Display`). For a self-hosted font, set `source` and `format` as well.'),
    weight: z
        .number()
        .optional()
        .describe('The font weight. Typically `400` for regular and `700` for bold. Variable fonts support precise weights like `550`.'),
    source: fontSourceSchema
        .optional()
        .describe('URL or path to a self-hosted font file. Use this when the font is not on Google Fonts. Example: `/fonts/custom.woff2` or `https://cdn.example.com/font.woff2`.'),
    format: z
        .enum(fontFormats)
        .optional()
        .describe('File format of the self-hosted font. Required when `source` is set.'),
})
    .refine((data) => {
    if (data.source) {
        return !!data.format;
    }
    return true;
});
export const fontsSchema = z
    .union([
    fontDetailsSchema,
    z
        .object({
        heading: fontDetailsSchema
            .optional()
            .describe('Font settings applied to all headings (`h1`–`h6`).'),
        body: fontDetailsSchema
            .optional()
            .describe('Font settings applied to body text and the rest of the page.'),
    })
        .strict(),
])
    .describe('Typography configuration. Use a single set of font settings for the whole site, or split into separate `heading` and `body` fonts.');
