// @mintlify/validation/dist/mint-config/schemas/v2/properties/background.js
import { z } from 'zod';
import { hexColor } from './reusable/color.js';
export const backgroundSchema = z
    .object({
    image: z
        .union([
        z
            .string()
            .describe('Path or URL pointing to the background image file, including the file extension. Used for both light and dark mode.'),
        z
            .object({
            light: z
                .string()
                .describe('The file for the light version of the background image used in dark mode, including the file extension.'),
            dark: z
                .string()
                .describe('The file for the dark version of the background image used in light mode, including the file extension.'),
        })
            .describe('The path to the background image. Can be a single file or a pair for light and dark mode.'),
    ])
        .optional()
        .describe('The background image configuration. Can be a single image for both light and dark mode, or separate images for each mode.'),
    decoration: z
        .enum(['gradient', 'grid', 'windows'])
        .optional()
        .describe('The background decoration of the theme'),
    color: z
        .object({
        light: hexColor.describe('The background color in light mode').optional(),
        dark: hexColor.describe('The background color in dark mode').optional(),
    })
        .strict()
        .optional()
        .describe('The colors of the background'),
})
    .describe('Background color and decoration settings');
